spv/sync.go

```diff
@@ -0,0 +1,1420 @@
+	}
+	for _, rp := range s.remotes {
+		err := f(rp)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Syncer) pickRemote(pick func(*p2p.RemotePeer) bool) (*p2p.RemotePeer, error) {
+	defer s.remotesMu.Unlock()
+	s.remotesMu.Lock()
+
+	for _, rp := range s.remotes {
+		if pick(rp) {
+			return rp, nil
+		}
+	}
+	return nil, errors.E(errors.NoPeers)
+}
+
+func (s *Syncer) getTransactionsByHashes(ctx context.Context, txHashes []*chainhash.Hash) ([]*wire.MsgTx, []*wire.InvVect, error) {
+	if len(txHashes) == 0 {
+		return nil, nil, nil
+	}
+
+	var notFound []*wire.InvVect
+	var foundTxs []*wire.MsgTx
+
+	for walletID, w := range s.wallets {
+		var missingWalletTxs []*wire.InvVect
+		var foundWalletTxs []*wire.MsgTx
+		var err error
+		foundWalletTxs, missingWalletTxs, err = w.GetTransactionsByHashes(ctx, txHashes)
+		if err != nil && !errors.Is(errors.NotExist, err) {
+			return nil, nil, errors.E("[%d] Failed to look up transactions for getdata reply to peer: %v", walletID, err)
+		}
+		if len(missingWalletTxs) != 0 {
+			notFound = append(notFound, missingWalletTxs...)
+		}
+
+		if len(foundWalletTxs) != 0 {
+			foundTxs = append(foundTxs, foundWalletTxs...)
+		}
+
+		for _, tx := range foundWalletTxs {
+			for index, hash := range txHashes {
+				if tx.TxHash() == *hash {
+					txHashes = append(txHashes[:index], txHashes[index+1:]...)
+					break
+				}
+			}
+
+			// delete the transactions that were not found by other wallets
+			for index, notFoundTx := range notFound {
+				if tx.TxHash() == notFoundTx.Hash {
+					notFound = append(notFound[:index], notFound[index+1:]...)
+					break
+				}
+			}
+		}
+
+		if len(txHashes) == 0 {
+			break
+		}
+	}
+
+	return foundTxs, notFound, nil
+}
+
+// receiveGetData handles all received getdata requests from peers.  An inv
+// message declaring knowledge of the data must have been previously sent to the
+// peer, or a notfound message reports the data as missing.  Only transactions
+// may be queried by a peer.
+func (s *Syncer) receiveGetData(ctx context.Context) error {
+	var wg sync.WaitGroup
+	for {
+		rp, msg, err := s.lp.ReceiveGetData(ctx)
+		if err != nil {
+			wg.Wait()
+			return err
+		}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			// Ensure that the data was (recently) announced using an inv.
+			var txHashes []*chainhash.Hash
+			var notFound []*wire.InvVect
+			for _, inv := range msg.InvList {
+				if !rp.InvsSent().Contains(inv.Hash) {
+					notFound = append(notFound, inv)
+					continue
+				}
+				switch inv.Type {
+				case wire.InvTypeTx:
+					txHashes = append(txHashes, &inv.Hash)
+				default:
+					notFound = append(notFound, inv)
+				}
+			}
+
+			// Search for requested transactions
+			var foundTxs []*wire.MsgTx
+			if len(txHashes) != 0 {
+				var missing []*wire.InvVect
+				var err error
+				foundTxs, missing, err = s.getTransactionsByHashes(ctx, txHashes)
+				if err != nil && !errors.Is(errors.NotExist, err) {
```

Comment:
`errors.Is` expects the first parameter to be the error being checked and the second parameter to be the target error.
```suggestion
				if err != nil && !errors.Is(err, errors.NotExist) {
```